### PR TITLE
Expose wakeupWorkflowInstance via REST API

### LINKED ISSUE #339: Expose `WorkflowInstanceServicewakeupWorkflowInstance` via REST API
```
  public boolean wakeupWorkflowInstance(long id, List<String> expectedStates) {
```

### PR DESCRIPTION
**Pull request by jsyrjala**
PUT /nflow/api/v1/workflow-instance/5/wakeup

Fixes #339

### CODE REVIEW

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java [hunk 1]
```diff
@@ -28,6 +28,7 @@
 import javax.ws.rs.QueryParam;
 import javax.ws.rs.core.Response;
 
+import io.nflow.rest.v1.msg.*;
```
Comment (efonsell) [inline nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java]: expand imports

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import javax.inject.Inject;
 import javax.validation.Valid;
 
+import io.nflow.rest.v1.msg.*;
```
Comment (efonsell) [inline nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java]: expand imports

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.nflow.engine.model.ModelObject;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(description = "Response to wake up request.")
+@SuppressFBWarnings(value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD", justification = "jackson reads dto fields")
+public class WakeupResponse extends ModelObject {
+    @ApiModelProperty("List of expected states. Can be empty, meaning any state.")
```
Comment (efonsell) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java]: Fix this property description

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.nflow.engine.model.ModelObject;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(description = "Response to wake up request.")
+@SuppressFBWarnings(value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD", justification = "jackson reads dto fields")
+public class WakeupResponse extends ModelObject {
+
+  @ApiModelProperty("List of expected states. Can be empty, meaning any state.")
```
Comment (efonsell) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java]: fix this property description

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import io.nflow.engine.model.ModelObject;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(description = "Response to wake up request.")
+@SuppressFBWarnings(value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD", justification = "jackson reads dto fields")
+public class WakeupResponse extends ModelObject {
+
+  @ApiModelProperty("True if the instance was woken up, false otherwise.")
+    public boolean wakeupSuccess;
```
Comment (efonsell) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java]: fix indentation

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java [hunk 6]
```diff
@@ -147,4 +151,16 @@ public Response setSignal(@ApiParam("Internal id for workflow instance") @PathPa
...
   }
 
+  @PUT
+  @Path("/{id}/wakeup")
+  @ApiOperation(value = "Wake up sleeping workflow instance. If expected states are given, only wake up if the instance is in one of the expected states.")
+  @ApiResponses({ @ApiResponse(code = 200, message = "When workflow wakeup was attempted")})
+  public WakeupResponse wakeup(@ApiParam("Internal id for workflow instance") @PathParam("id") int id,
+      @Valid @ApiParam("Expected states") WakeupRequest req) {
+    WakeupResponse response = new WakeupResponse();
+    List<String> expectedStates = req.expectedStates != null ? req.expectedStates : emptyList();
```
Comment (efonsell) [inline nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java]:
```suggestion
    List<String> expectedStates = Optional.ofNullable(req.expectedStates).orElseGet(Collections::emptyList);
```

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java [hunk 7]
```diff
@@ -128,4 +131,15 @@ public WorkflowInstanceResource(WorkflowInstanceService workflowInstances, Creat
     return (updated ? ResponseEntity.ok("Signal was set successfully") : ResponseEntity.ok("Signal was not set"));
   }
 
+  @PutMapping(path = "/{id}/wakeup", consumes = APPLICATION_JSON_VALUE)
+  @ApiOperation(value = "Wake up sleeping workflow instance. If expected states are given, only wake up if the instance is in one of the expected states.")
+  @ApiResponses({ @ApiResponse(code = 200, message = "When workflow wakeup was attempted")})
+  public WakeupResponse wakeup(@ApiParam("Internal id for workflow instance") @PathVariable("id") int id,
+      @RequestBody @Valid @ApiParam("Expected states") WakeupRequest req) {
+    WakeupResponse response = new WakeupResponse();
+    List<String> expectedStates = req.expectedStates != null ? req.expectedStates : emptyList();
```
Comment (efonsell) [inline nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java]:
```suggestion
    List<String> expectedStates = Optional.ofNullable(req.expectedStates).orElseGet(Collections::emptyList);
```

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java (hunk 6 above)
Comment (jsyrjala) [reply nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java]: Fixed. Not sure though if it is better...

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java (hunk 7 above)
Comment (jsyrjala) [reply nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java]: fixed

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java (hunk 5 above)
Comment (jsyrjala) [reply nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/WakeupResponse.java]: fxied

**Pull request merged**